lib/diagnostics_channel.js

```diff
@@ -138,10 +226,171 @@ function hasSubscribers(name) {
   return channel.hasSubscribers;
 }
 
+const traceEvents = [
+  'start',
+  'end',
+  'asyncStart',
+  'asyncEnd',
+  'error',
+];
+
+function assertChannel(value, name) {
+  if (!(value instanceof Channel)) {
+    throw new ERR_INVALID_ARG_TYPE(name, ['Channel'], value);
+  }
+}
+
+class TracingChannel {
+  constructor(nameOrChannels) {
+    if (typeof nameOrChannels === 'string') {
+      this.start = channel(`tracing:${nameOrChannels}:start`);
+      this.end = channel(`tracing:${nameOrChannels}:end`);
+      this.asyncStart = channel(`tracing:${nameOrChannels}:asyncStart`);
+      this.asyncEnd = channel(`tracing:${nameOrChannels}:asyncEnd`);
+      this.error = channel(`tracing:${nameOrChannels}:error`);
+    } else if (typeof nameOrChannels === 'object') {
+      const { start, end, asyncStart, asyncEnd, error } = nameOrChannels;
+
+      assertChannel(start, 'nameOrChannels.start');
+      assertChannel(end, 'nameOrChannels.end');
+      assertChannel(asyncStart, 'nameOrChannels.asyncStart');
+      assertChannel(asyncEnd, 'nameOrChannels.asyncEnd');
+      assertChannel(error, 'nameOrChannels.error');
+
+      this.start = start;
+      this.end = end;
+      this.asyncStart = asyncStart;
+      this.asyncEnd = asyncEnd;
+      this.error = error;
+    } else {
+      throw new ERR_INVALID_ARG_TYPE('nameOrChannels',
+                                     ['string', 'object', 'Channel'],
+                                     nameOrChannels);
+    }
+  }
+
+  subscribe(handlers) {
+    for (const name of traceEvents) {
+      if (!handlers[name]) continue;
+
+      this[name]?.subscribe(handlers[name]);
+    }
+  }
+
+  unsubscribe(handlers) {
+    let done = true;
+
+    for (const name of traceEvents) {
+      if (!handlers[name]) continue;
+
+      if (!this[name]?.unsubscribe(handlers[name])) {
+        done = false;
+      }
+    }
+
+    return done;
+  }
+
+  traceSync(fn, ctx = {}, thisArg, ...args) {
+    const { start, end, error } = this;
+
+    try {
+      const result = start.runStores(ctx, fn, thisArg, ...args);
+      ctx.result = result;
+      return result;
+    } catch (err) {
+      ctx.error = err;
+      error.publish(ctx);
+      throw err;
+    } finally {
+      end.publish(ctx);
+    }
+  }
+
+  tracePromise(fn, ctx = {}, thisArg, ...args) {
+    const { start, end, asyncStart, asyncEnd, error } = this;
+
+    function reject(err) {
+      ctx.error = err;
+      error.publish(ctx);
+      asyncStart.publish(ctx);
+      // TODO: Is there a way to have asyncEnd _after_ the continuation?
+      asyncEnd.publish(ctx);
+      return PromiseReject(err);
+    }
+
+    function resolve(result) {
+      ctx.result = result;
+      asyncStart.publish(ctx);
```

Comment:
`asyncStart` should also use `runStore` so that the parent context from when `bindStore` was called can be restored.

Reply:
There's no way to do that around a promise resolve. What would be the function to scope the `runStore`?

Reply:
I've replaced all the tracing channel publishes with `runStores` so transformers can be used to rebind as-needed. I'm not sure I like the approach though. 😐

Reply:
What is the use case behind transforming the ctx several times per traceable operation?
Would be nice to have this somewhere shown in a test.

Reply:
I'm not convinced it's actually necessary, but it could be used to restore the context object created in the start channel for any subsequent event publishes in the event the context was lost due to some other issue internally. It can definitely be helpful to allow APMs to have some failure recovery in the event of context loss but it _is_ kind of overlooking the _actual_ issue of ensuring context loss doesn't happen in the first place. 🤷🏻

Reply:
Right now, at least for APM vendors, `ctx` would be used for the following 3 things:

1. Store the span of the operation represented by the tracing channel so that it can be retrieved in `asyncStart` since it would be bound to a different store.
2. Pass the error to the `error` event while still being able to pass `ctx`. Since only a single argument is allowed, one needs to go on the other to pass both.
3. Pass the result to the `asyncStart` event, for the same reason as above for errors.

I think maybe it could be avoided entirely, but it would require a lot more work on the subscriber side, and it could end up being a lot less efficient and prone to objects lingering in memory.

For example, in order to avoid any updates to `ctx,` the following would be needed on the subscriber side:

1. Enter a new store and add the current store to it with `store.parentStore = storage.getStore()` in `start`.
2. Publish the error instead of `ctx` on the `error` channel.
3. Publish the result instead of `ctx` on the `end` or `asyncStart` channel. This would have the side effect however of not allowing `end` to ever be passed anything because the presence of a result would be used to indicate that the operation was synchronous. This could be mitigated by always passing an object with or without a result on it, or by changing the API to only include a single `end` instead of one for sync and one for async. Both would be problematic if a subscriber is interested in something from both `end` and `asyncStart`.
4. Restore the parent store from `store.parentStore` and then delete `store.parentStore` to avoid memory leaks in `asyncStart` subscriber.

So basically, updating `ctx` is not necessarily _needed_, but not updating it would remove its usefulness completely in the first place and would put a lot of burden on subscribers. It would also make `runStores` mostly useless too since at that point the subscribers would have to do enough work that it might as well also handle storage. As @Qard said, it would also make it impossible to recover when the context is lost. It doesn't mean it shouldn't be considered, but it's important to understand the trade-offs.

Reply:
Related: https://github.com/nodejs/node/pull/44943#discussion_r1100931452

Reply:
The data published within the TracingChannel APIs is always `context` (named `data` in the docs) so how can this get lost within e.g. `traceCallback`?

Or do you refer to the async context tracked (hopefully) by AsyncLocalStorage?

Reply:
The `ctx` is only meant for sharing data between the events of that particular trace call. If you want to share data beyond that, such as across a request trace, you should use bindStore/runStores to feed that data into AsyncLocalStorage, making that the state of the store within any descending async activity and therefore the state when any of the `async*` events happen. Yes, there is a risk of context loss, which is why I stated using runStores around the callback _could_ be used to restore the state where necessary. But async context correctness is not and _should not_ be a responsibility of this API. Context loss bugs should be fixed where the actual context loss is and not here.

The reason a shared object is used between the events is to enable sharing data between those events. That could be sharing a span object by storing it on the `ctx` or it could just mean a shared object identity to use as a weakmap key. In either case, it is _only_ designed for sharing within the trace call itself. I do not intend for any linkage between separate traces to be part of the API as that should be left entirely up to the consumer. The parent to child relationship you want should be done on the user side with `bindStore`:

```js
channel.bindStore(spanStore, (data) => {
  const parentSpan = spanStore.getStore()
  return new Span(data, parentSpan)
})
```

Reply:
I'm ok with using `runStores` for `asyncStart` in the `traceCallback` but I think we should clearly document this and also verify it in tests.
I think it's quite special that  `asyncStart` uses `runStores` in `traceCallback` but not in `tracePromise`.
Also that `runStores` is not used for `error` channel in the callback nor the `asyncEnd` channel uses it (runs within `asyncStart` instead).